src/foam/dao/JDAO.java

```diff
@@ -7,24 +7,117 @@
 package foam.dao;
 
 import foam.core.ClassInfo;
-import foam.lib.json.Outputter;
-import foam.lib.json.OutputterMode;
+import foam.core.FObject;
+import foam.core.X;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+import foam.nanos.auth.User;
+import foam.util.SafetyUtil;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.TimeZone;
 
 public class JDAO
-  extends AbstractJDAO
+  extends ProxyDAO
 {
-  protected Outputter outputter_ = new Outputter(OutputterMode.STORAGE);
+  protected static ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder b = super.get();
+      b.setLength(0);
+      return b;
+    }
+  };
+
+  protected static final ThreadLocal<SimpleDateFormat> sdf = new ThreadLocal<SimpleDateFormat>() {
+    @Override
+    protected SimpleDateFormat initialValue() {
+      SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
+      sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
+      return sdf;
+    }
+  };
 
-  public JDAO(foam.core.X x, ClassInfo classInfo, String filename) {
+  protected FileJournal journal_;
+
+  public JDAO(X x, ClassInfo classInfo, String filename) {
     this(x, new MapDAO(classInfo), filename);
   }
 
-  public JDAO(foam.core.X x, DAO delegate, String filename) {
-    super(x, delegate, filename);
+  public JDAO(X x, DAO delegate, String filename) {
+    setX(x);
+    setOf(delegate.getOf());
+    setDelegate(delegate);
+
+    // create journal
+    journal_ = new FileJournal.Builder(getX())
+      .setDao(delegate)
+      .setFilename(filename)
+      .setCreateFile(true)
+      .build();
+
+    // create a composite journal of repo journal
+    // and runtime journal and load them all
+    new CompositeJournal.Builder(getX())
+      .setDelegates(new Journal[]{
+        new FileJournal.Builder(getX())
+          .setFilename(filename + ".0")
+          .build(),
+        new FileJournal.Builder(getX())
+          .setFilename(filename)
+          .build()
+      })
+      .build().replay(delegate);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    writeComment((User) x.get("user"));
```

Comment:
Could this writeComment stuff be a journal decorator too?

Reply:
I would have liked it to have been but the Sink interface doesn’t provide the context anywhere for me to get the current user.